service/history/queue/timer_queue_processor_base.go

```diff
@@ -0,0 +1,590 @@
+		taskProcessor:   taskProcessor,
+		redispatchQueue: redispatchQueue,
+		timerGate:       timerGate,
+
+		options:               options,
+		updateMaxReadLevel:    updateMaxReadLevel,
+		updateClusterAckLevel: updateClusterAckLevel,
+		queueShutdown:         queueShutdown,
+		taskInitializer:       taskInitializer,
+
+		logger:        logger.WithTags(tag.ComponentTimerQueue),
+		metricsClient: metricsClient,
+		metricsScope:  metricsClient.Scope(options.MetricScope),
+
+		rateLimiter: quotas.NewDynamicRateLimiter(
+			func() float64 {
+				return float64(options.MaxPollRPS())
+			},
+		),
+		lastPollTime: time.Time{},
+		status:       common.DaemonStatusInitialized,
+		shutdownCh:   make(chan struct{}),
+		newTimerCh:   make(chan struct{}, 1),
+
+		processingQueueCollections: newProcessingQueueCollections(
+			processingQueueStates,
+			logger,
+			metricsClient,
+		),
+		processingQueueReadProgress: make(map[int]timeTaskReadProgress),
+	}
+}
+
+func (t *timerQueueProcessorBase) Start() {
+	if !atomic.CompareAndSwapInt32(&t.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	t.logger.Info("", tag.LifeCycleStarting)
+	defer t.logger.Info("", tag.LifeCycleStarted)
+
+	t.notifyNewTimer(time.Time{})
+
+	t.shutdownWG.Add(1)
+	go t.processorPump()
+}
+
+func (t *timerQueueProcessorBase) Stop() {
+	if !atomic.CompareAndSwapInt32(&t.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	t.logger.Info("", tag.LifeCycleStopping)
+	defer t.logger.Info("", tag.LifeCycleStopped)
+
+	t.timerGate.Close()
+	close(t.shutdownCh)
+
+	if success := common.AwaitWaitGroup(&t.shutdownWG, time.Minute); !success {
+		t.logger.Warn("", tag.LifeCycleStopTimedout)
+	}
+}
+
+func (t *timerQueueProcessorBase) processorPump() {
+	defer t.shutdownWG.Done()
+
+	pollTimer := time.NewTimer(backoff.JitDuration(
+		t.options.MaxPollInterval(),
+		t.options.MaxPollIntervalJitterCoefficient(),
+	))
+	defer pollTimer.Stop()
+
+	updateAckTimer := time.NewTimer(backoff.JitDuration(
+		t.options.UpdateAckInterval(),
+		t.options.UpdateAckIntervalJitterCoefficient(),
+	))
+	defer updateAckTimer.Stop()
+
+	redispatchTimer := time.NewTimer(backoff.JitDuration(
+		t.options.RedispatchInterval(),
+		t.options.RedispatchIntervalJitterCoefficient(),
+	))
+	defer redispatchTimer.Stop()
+
+	splitQueueTimer := time.NewTimer(backoff.JitDuration(
+		t.options.SplitQueueInterval(),
+		t.options.SplitQueueIntervalJitterCoefficient(),
+	))
+	defer splitQueueTimer.Stop()
+
+processorPumpLoop:
+	for {
+		select {
+		case <-t.shutdownCh:
+			break processorPumpLoop
+		case <-t.timerGate.FireChan():
+			if t.redispatchQueue.Len() <= t.options.MaxRedispatchQueueSize() {
+				t.processBatch()
+				continue
+			}
+
+			RedispatchTasks(
+				t.redispatchQueue,
+				t.taskProcessor,
+				t.logger,
+				t.metricsScope,
+				t.shutdownCh,
+			)
+			t.notifyNewTimer(time.Time{})
+		case <-pollTimer.C:
+			pollTimer.Reset(backoff.JitDuration(
```

Comment:
should those timer resets happen after the actions?

Reply:
Yes, makes sense. I will also update the transfer queue processor.

Reply:
Updated.